### PR TITLE
aditi.maheshwari/telemetry app product change test

### PR DESCRIPTION
**Pull request by aditim16**
## Description

Assert that this event app-product-change is emitted when a product is enabled 

## Workflow

1. ⚠️⚠️ Create your PR as draft (we're receiving lot of PR, it saves us lot of time) ⚠️⚠️
2. Follows the style guidelines of this project (See [how to easily lint the code](https://github.com/DataDog/system-tests/blob/main/docs/edit/lint.md))
3. Work on you PR until the CI passes (if something not related to your task is failing, you can ignore it)
4. Mark it as ready for review

> **_NOTE:_**  By default in PR only default scenario tests will be launched. Please refer to the [documentation](https://datadoghq.atlassian.net/wiki/spaces/APMINT/pages/2866381467/CI+Workflow+Github+Actions) to run all scenarios in your PR if needed.

Once your PR is reviewed, you can merge it ! :heart:

### CODE REVIEW

**Review by cbeauchesne — CHANGES_REQUESTED** [review]

File: tests/test_telemetry.py [hunk 1]
```diff
@@ -359,3 +359,32 @@ def test_app_dependencies_loaded(self):
...
+                    assert appsecEnabled == True, f"Product appsec enabled flag is {appsecEnabled}"
+                    profilerEnabled = prod["profiler"]["enabled"]
+                    assert profilerEnabled == True, f"Product appsec enabled flag is {profilerEnabled}"
+                    dynamicInstrumentationEnabled = prod["dynamic_instrumentation"]["enabled"]
+                    assert (
+                        dynamicInstrumentationEnabled == False
+                    ), f"Product appsec enabled flag is {dynamicInstrumentationEnabled}"
+            else:
+                raise Exception("app-product-change is not emited when product change is enabled")
+            self.validate_library_telemetry_data(validator)
```
[the remark this answers is not in the document]
Comment (aditim16) [reply tests/test_telemetry.py]: fixed it

File: tests/test_telemetry.py [hunk 2]
```diff
@@ -359,3 +359,47 @@ def test_app_dependencies_loaded(self):
         for dependency, seen in seen_loaded_dependencies.items():
             if not seen:
                 raise Exception(dependency + " not recieved in app-dependencies-loaded message")
+
+    def setup_app_product_change(self):
+        weblog.get("/enable_product")
+
+    @bug(
```
Comment (cbeauchesne) [inline tests/test_telemetry.py:366]: When an endpoint is missing, we categorize it as missing feature.

```suggestion
    @missing_feature(context.library in ("dotnet", "nodejs", ...),   reason="weblog GET/enable_product and app-product-change event is not implemented yet.")
```

**Review by cbeauchesne — CHANGES_REQUESTED** [review]

File: tests/test_telemetry.py [hunk 3]
```diff
@@ -369,3 +369,31 @@ def test_app_dependencies_loaded(self):
         for dependency, seen in seen_loaded_dependencies.items():
             if not seen:
                 raise Exception(dependency + " not recieved in app-dependencies-loaded message")
+
+    def setup_app_product_change(self):
+        weblog.get("/enable_product")
+
+    @missing_feature(
+        context.library in ("dotnet", "nodejs", "java", "python"),
```
Comment (cbeauchesne) [inline tests/test_telemetry.py:377]: All of them, as no tracer implemented yet this feature

File: tests/test_telemetry.py [hunk 4]
```diff
@@ -369,3 +369,31 @@ def test_app_dependencies_loaded(self):
...
+        reason="Weblog GET/enable_product and app-product-change event is not implemented yet.",
+    )
+    def test_app_product_change(self):
+        """Test product change data when product is enabled"""
+
+        def validator(data):
+            content = data["request"]["content"]
+            if content.get("request_type") == "app-product-change":
+                products = content["payload"]["products"]
+                for prod in products > 0:
```
Comment (cbeauchesne) [inline tests/test_telemetry.py:387]:
```suggestion
                for prod in products:
```

File: tests/test_telemetry.py [hunk 5]
```diff
@@ -369,3 +369,31 @@ def test_app_dependencies_loaded(self):
...
+                    appsecEnabled = prod["appsec"]["enabled"]
+                    assert appsecEnabled == True, f"Product appsec enabled flag is {appsecEnabled}"
```
Comment (cbeauchesne) [inline tests/test_telemetry.py:389]:
```suggestion
                    assert appsecEnabled is True, f"Product appsec enabled flag is {appsecEnabled}"
```

File: tests/test_telemetry.py [hunk 6]
```diff
@@ -369,3 +369,31 @@ def test_app_dependencies_loaded(self):
...
+                    profilerEnabled = prod["profiler"]["enabled"]
+                    assert profilerEnabled == True, f"Product appsec enabled flag is {profilerEnabled}"
+                    dynamicInstrumentationEnabled = prod["dynamic_instrumentation"]["enabled"]
+                    assert (
+                        dynamicInstrumentationEnabled == False
+                    ), f"Product appsec enabled flag is {dynamicInstrumentationEnabled}"
```
Comment (cbeauchesne) [inline tests/test_telemetry.py:395]: better error message is like "dynamic_instrumentation was expected to be False, found True"

File: tests/test_telemetry.py [hunk 7]
```diff
@@ -369,3 +369,31 @@ def test_app_dependencies_loaded(self):
         for dependency, seen in seen_loaded_dependencies.items():
             if not seen:
                 raise Exception(dependency + " not recieved in app-dependencies-loaded message")
+
+    def setup_app_product_change(self):
```
Comment (cbeauchesne) [inline tests/test_telemetry.py:373]: Can you do a dedicated scenario with appsec, profiling ... set to false at startup time.

File: tests/test_telemetry.py [hunk 8]
```diff
@@ -369,3 +369,31 @@ def test_app_dependencies_loaded(self):
...
+                for prod in products > 0:
+                    appsecEnabled = prod["appsec"]["enabled"]
+                    assert appsecEnabled == True, f"Product appsec enabled flag is {appsecEnabled}"
+                    profilerEnabled = prod["profiler"]["enabled"]
+                    assert profilerEnabled == True, f"Product appsec enabled flag is {profilerEnabled}"
+                    dynamicInstrumentationEnabled = prod["dynamic_instrumentation"]["enabled"]
+                    assert (
+                        dynamicInstrumentationEnabled == False
+                    ), f"Product appsec enabled flag is {dynamicInstrumentationEnabled}"
+            else:
```
Comment (cbeauchesne) [inline tests/test_telemetry.py:396]: Remove this part, and add another line 

```
    self.validate_library_telemetry_data(lambda data: data["request"]["content"].get("request_type") == "app-product-change")
```

**Review by cbeauchesne — CHANGES_REQUESTED** [review]

**Comment by cbeauchesne** [comment]
Could you resolve the conflict before I review it please ? 🙏

File: tests/test_telemetry.py [hunk 9]
```diff
@@ -370,6 +370,65 @@ def test_app_dependencies_loaded(self):
...
+                    appsec_enabled = prod["appsec"]["enabled"]
+                    assert (
+                        appsec_enabled is True
+                    ), f"Product appsec Product profiler enabled was expected to be True, found False"
+                    profiler_enabled = prod["profiler"]["enabled"]
+                    assert profiler_enabled is True, f"Product profiler enabled was expected to be True, found False"
+                    dynamic_instrumentation_enabled = prod["dynamic_instrumentation"]["enabled"]
+                    assert (
+                        dynamic_instrumentation_enabled is False
+                    ), f"Product dynamic_instrumentation enabled was expected to be False, found True"
```
Comment (mehulsonowal) [inline tests/test_telemetry.py:403]:
```suggestion
                for product in products:
                    appsec_enabled = product["appsec"]["enabled"]
                    profiler_enabled = product["profiler"]["enabled"]
                    dynamic_instrumentation_enabled = product["dynamic_instrumentation"]["enabled"]
                    assert (
                        appsec_enabled is True
                    ), f"Product appsec Product profiler enabled was expected to be True, found False"
                    assert profiler_enabled is True, f"Product profiler enabled was expected to be True, found False"
                    assert (
                        dynamic_instrumentation_enabled is False
                    ), f"Product dynamic_instrumentation enabled was expected to be False, found True"
```

**Review by mehulsonowal — APPROVED** [review]
small nit but LGTM!

File: .github/workflows/ci.yml [hunk 10]
```diff
@@ -364,6 +364,9 @@ jobs:
       run: ./run.sh SAMPLING
       env:
         DD_API_KEY: ${{ secrets.DD_API_KEY }}
+    - name: Run TELEMETRY_APP_STARTED_PRODUCTS_DISABLED scenario
+      if: steps.build.outcome == 'success' && github.event.action != 'opened' && !contains(github.event.pull_request.labels.*.name, 'run-default-scenario') 
+      run: ./run.sh TELEMETRY_APP_STARTED_PRODUCTS_DISABLED
     - name: Run TELEMETRY_DEPENDENCY_LOADED_TEST_FOR_DEPENDENCY_COLLECTION_DISABLED scenario
```
Comment (cbeauchesne) [inline .github/workflows/ci.yml:370]: `env` is missing

**Review by cbeauchesne — CHANGES_REQUESTED** [review]

**Comment by cbeauchesne** [comment]
Could you resolve the conflict please ? :)

**Comment by cbeauchesne** [comment]
CI is failing on `spring-boot-3-native` and `spring-boot-native`, if it's expected, can add the in `missing_feature` please ?

**Comment by cbeauchesne** [comment]
CI is still failing on java native 3

**Review by cbeauchesne — APPROVED** [review]

**Pull request merged**